### PR TITLE
feat(core): add file-based session persistence (M1)

### PR DESCRIPTION
**Pull request by cidrblock**
## Summary

- Implement `SessionStore` class with file-based JSON storage and index for fast listing (DR-021)
- Wire session persistence into all three transports: gRPC handlers, web REST API, and CLI
- Add `aby sessions list/show/delete` commands and `aby chat --session <id|new>` flag
- 19 unit tests + 11 integration tests, all passing

## Commits

- `feat(core): add file-based session persistence (M1)` — Full implementation of session CRUD, session-scoped chat, and documentation updates
- `fix(session): address Copilot review feedback on session persistence` — Disable tools in session chat (no approval mechanism), add async write lock to SessionStore, reuse state.sessionStore in CLI, validate pagination params, differentiate 404/500, harden readIndex error handling

## Key changes

- **`core/paths.ts`**: Add `getDataDir()` (`$XDG_DATA_HOME/abbenay`) and `getSessionsDir()`
- **`core/session-store.ts`** (new): `SessionStore` class with `create`, `get`, `list`, `delete`, `appendMessage`, `updateTitle` — atomic writes via tmp+rename, `index.json` for O(1) listing, async write lock for concurrent mutation safety
- **`daemon/state.ts`**: Wire `SessionStore` instance into `DaemonState`
- **`daemon/server/abbenay-service.ts`**: Replace 5 session stub RPCs with real implementations (`CreateSession`, `GetSession`, `ListSessions`, `DeleteSession`, `SessionChat`)
- **`daemon/web/server.ts`**: Add `POST/GET/DELETE /api/sessions`, `POST /api/sessions/:id/chat` with SSE streaming + message persistence + auto-titling
- **`daemon/index.ts`**: Add `aby sessions` command group (list/show/delete) and `--session` flag on `aby chat`
- **`daemon/chat.ts`**: Support `--session <id>` for resuming and `--session new` for creating persistent chat sessions
- **`docs/decisions.md`**: Add DR-021 (file-based session storage rationale)
- **`docs/ROADMAP.md`**: Mark Phase 3 M1+M3 as done
- **`docs/ARCHITECTURE.md`**: Add session-store to core table, update session management section, file locations
- **`docs/TESTING.md`**: Add new test files to tree and tables

## Test plan

- [x] `npm run lint` passes locally (0 errors)
- [x] `npm test` passes locally (263 tests, 13 suites)
- [x] TypeScript compilation passes (`tsc --noEmit`)
- [x] New `session-store.test.ts` — 19 tests covering CRUD, appendMessage, updateTitle, index consistency, concurrent writes, edge cases
- [x] New `tests/integration/sessions.test.ts` — 11 tests covering all REST endpoints, session chat SSE streaming + persistence, auto-titling, error cases
- [x] Existing test suites unaffected (all 263 tests pass)

### CODE REVIEW

File: packages/daemon/src/daemon/web/server.ts [hunk 1]
```diff
@@ -892,6 +892,135 @@ export function createWebApp(state: DaemonState): Express {
...
+          if (ended) break;
+          if (chunk.type === 'text' && chunk.text) {
+            fullText += chunk.text;
+            safeWrite(`data: ${JSON.stringify({ type: 'text', content: chunk.text })}\n\n`);
+          } else if (chunk.type === 'error') {
+            safeWrite(`data: ${JSON.stringify({ type: 'error', error: chunk.error })}\n\n`);
+          } else if (chunk.type === 'done') {
+            safeWrite(`data: ${JSON.stringify({ type: 'done', finish_reason: chunk.finishReason || 'stop' })}\n\n`);
+          }
+        }
```
[the remark this answers is not in the document]
Comment (cidrblock) [reply packages/daemon/src/daemon/web/server.ts]: Fixed in 5fcaabe — session chat endpoints (web + gRPC) now default toolMode to 'none'. Tools are disabled until an approval mechanism is wired into these transports.

File: packages/daemon/src/core/session-store.ts [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+    const session = await this.get(id, true);
+    session.title = title;
+    session.updatedAt = new Date().toISOString();
+
+    await this.writeSession(session);
+    await this.updateIndex(id, { title, updatedAt: session.updatedAt });
+  }
+
+  // ── Private helpers ───────────────────────────────────────────────────
+
```
[the remark this answers is not in the document]
Comment (cidrblock) [reply packages/daemon/src/core/session-store.ts]: Fixed in 5fcaabe — added a promise-chain async mutex (withWriteLock) to SessionStore. All mutating methods (create, delete, appendMessage, updateTitle) are now serialized.

File: packages/daemon/src/daemon/chat.ts [hunk 3]
```diff
@@ -44,6 +47,37 @@ export async function runInteractiveChat(options: ChatOptions): Promise<void> {
 
   await state.initMcpConnections();
 
+  // Resolve session if requested
+  let sessionId: string | undefined;
+  let model = options.model || '';
+  const store = new SessionStore(getSessionsDir());
```
[the remark this answers is not in the document]
Comment (cidrblock) [reply packages/daemon/src/daemon/chat.ts]: Fixed in 5fcaabe — CLI chat now uses state.sessionStore instead of constructing a separate instance. Removed unused getSessionsDir import.

File: packages/daemon/src/daemon/server/abbenay-service.ts [hunk 4]
```diff
@@ -798,26 +830,148 @@ export function createAbbenayService(state: DaemonState) {
...
+        try {
+          const session = await state.sessionStore.get(sessionId, true);
+          await state.sessionStore.appendMessage(sessionId, chatMessage);
+
+          const allMessages = [...session.messages, chatMessage];
+          const toolOptions: ChatToolOptions = { toolMode, maxToolIterations };
+
+          let fullText = '';
+          for await (const chunk of state.chat(session.model, allMessages, hasParams ? requestParams : undefined, toolOptions)) {
+            if (chunk.type === 'text' && chunk.text) {
```
[the remark this answers is not in the document]
Comment (cidrblock) [reply packages/daemon/src/daemon/server/abbenay-service.ts]: Fixed in 5fcaabe — gRPC SessionChat now defaults toolMode to 'none'.

File: packages/daemon/src/core/session-store.ts [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+    sessions.sort((a, b) => b.updatedAt.localeCompare(a.updatedAt));
+
+    if (options?.model) {
+      sessions = sessions.filter((s) => s.model === options.model);
+    }
+
+    const totalCount = sessions.length;
+    const offset = options?.offset || 0;
+    const limit = options?.limit || sessions.length;
+
```
[the remark this answers is not in the document]
Comment (cidrblock) [reply packages/daemon/src/core/session-store.ts]: Fixed in 5fcaabe — list() now uses ?? for defaults and validates that offset/limit are finite non-negative integers, throwing on invalid values.

File: packages/daemon/src/daemon/web/server.ts [hunk 6]
```diff
@@ -892,6 +892,135 @@ export function createWebApp(state: DaemonState): Express {
...
+
+  app.get('/api/sessions/:id', async (req, res) => {
+    try {
+      const includeMessages = req.query.includeMessages !== 'false';
+      const session = await state.sessionStore.get(req.params.id, includeMessages);
+      res.json(session);
+    } catch (err: unknown) {
+      const msg = err instanceof Error ? err.message : String(err);
+      res.status(404).json({ error: msg });
+    }
```
[the remark this answers is not in the document]
Comment (cidrblock) [reply packages/daemon/src/daemon/web/server.ts:957]: Fixed in 5fcaabe — GET and DELETE /api/sessions/:id now check for 'not found' in the error message and return 500 with logging for other errors.

File: packages/daemon/src/daemon/web/server.ts [hunk 7]
```diff
@@ -892,6 +892,135 @@ export function createWebApp(state: DaemonState): Express {
...
+      console.error('[Web] POST /api/sessions error:', msg);
+      res.status(500).json({ error: msg });
+    }
+  });
+
+  app.get('/api/sessions', async (req, res) => {
+    try {
+      const model = req.query.model as string | undefined;
+      const limit = req.query.limit ? Number(req.query.limit) : undefined;
+      const offset = req.query.offset ? Number(req.query.offset) : undefined;
```
[the remark this answers is not in the document]
Comment (cidrblock) [reply packages/daemon/src/daemon/web/server.ts]: Fixed in 5fcaabe — web list endpoint now validates limit/offset as non-negative integers and returns 400 on invalid values before reaching SessionStore.

File: packages/daemon/src/core/session-store.ts [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+    await fs.promises.writeFile(tmpPath, JSON.stringify(session, null, 2));
+    await fs.promises.rename(tmpPath, filePath);
+  }
+
+  private async readIndex(): Promise<IndexFile> {
+    try {
+      const raw = await fs.promises.readFile(this.indexPath(), 'utf-8');
+      return JSON.parse(raw) as IndexFile;
+    } catch {
+      return { sessions: [] };
```
[the remark this answers is not in the document]
Comment (cidrblock) [reply packages/daemon/src/core/session-store.ts]: Fixed in 5fcaabe — readIndex() now only catches ENOENT (empty index). Parse errors, permission errors, and other I/O failures are rethrown.

File: packages/daemon/src/core/session-store.ts [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+      raw = await fs.promises.readFile(filePath, 'utf-8');
+    } catch {
+      throw new Error(`Session not found: ${id}`);
+    }
+
+    const session = JSON.parse(raw) as Session;
+    if (!includeMessages) {
+      return { ...session, messages: [] };
+    }
+    return session;
```
[the remark this answers is not in the document]
Comment (cidrblock) [reply packages/daemon/src/core/session-store.ts:119]: Acknowledged — the full-file read is acceptable for now since session files are small (<1MB). Will optimize with index-based metadata retrieval when sessions grow larger.

File: packages/daemon/src/daemon/server/abbenay-service.ts [hunk 10]
```diff
@@ -798,26 +830,148 @@ export function createAbbenayService(state: DaemonState) {
...
-    GetSession(call: grpc.ServerUnaryCall<object, object>, callback: grpc.sendUnaryData<object>): void {
-      callback({ code: grpc.status.UNIMPLEMENTED, message: 'Sessions not yet implemented' });
+
+    ListSessions(
+      call: grpc.ServerUnaryCall<ListSessionsRequestProto, object>,
+      callback: grpc.sendUnaryData<object>,
+    ): void {
+      const model = call.request.model_filter || call.request.modelFilter || undefined;
+      const limit = call.request.limit || undefined;
+      const offset = call.request.offset || undefined;
```
[the remark this answers is not in the document]
Comment (cidrblock) [reply packages/daemon/src/daemon/server/abbenay-service.ts]: Fixed in 5fcaabe — gRPC ListSessions now uses null-safe coalescing: treats null/undefined and negative values as unset, preserving explicit 0.
